rcl/test/rcl/test_info_by_topic.cpp

```diff
@@ -0,0 +1,378 @@
+
+#ifndef _WIN32
+# pragma GCC diagnostic ignored "-Wmissing-field-initializers"
+#endif
+
+#include <gtest/gtest.h>
+
+#include <memory>
+#include <string>
+
+#include "rcl/error_handling.h"
+#include "rcl/graph.h"
+#include "rcl/rcl.h"
+
+#include "rmw/topic_info_array.h"
+#include "rmw/error_handling.h"
+
+#include "test_msgs/msg/strings.h"
+#include "rosidl_generator_c/string_functions.h"
+
+#include "osrf_testing_tools_cpp/scope_exit.hpp"
+
+#ifdef RMW_IMPLEMENTATION
+# define CLASSNAME_(NAME, SUFFIX) NAME ## __ ## SUFFIX
+# define CLASSNAME(NAME, SUFFIX) CLASSNAME_(NAME, SUFFIX)
+#else
+# define CLASSNAME(NAME, SUFFIX) NAME
+#endif
+
+class CLASSNAME (TestInfoByTopicFixture, RMW_IMPLEMENTATION) : public ::testing::Test
+{
+public:
+  rcl_context_t * old_context_ptr;
+  rcl_context_t * context_ptr;
+  rcl_node_t * old_node_ptr;
+  rcl_node_t * node_ptr;
+  const char * test_graph_node_name = "test_graph_node";
+  std::unique_ptr<rmw_topic_info_array_t> topic_info_array;
+  const char * const topic_name = "valid_topic_name";
+  bool is_fastrtps;
+  void SetUp()
+  {
+    is_fastrtps = (std::string(rmw_get_implementation_identifier()).find("rmw_fastrtps") == 0);
+    rcl_ret_t ret;
+    rcl_init_options_t init_options = rcl_get_zero_initialized_init_options();
+    ret = rcl_init_options_init(&init_options, rcl_get_default_allocator());
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT({
+      EXPECT_EQ(RCL_RET_OK, rcl_init_options_fini(&init_options)) << rcl_get_error_string().str;
+    });
+    this->old_context_ptr = new rcl_context_t;
+    *this->old_context_ptr = rcl_get_zero_initialized_context();
+    ret = rcl_init(0, nullptr, &init_options, this->old_context_ptr);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    this->old_node_ptr = new rcl_node_t;
+    *this->old_node_ptr = rcl_get_zero_initialized_node();
+    const char * old_name = "old_node_name";
+    rcl_node_options_t node_options = rcl_node_get_default_options();
+    ret = rcl_node_init(this->old_node_ptr, old_name, "", this->old_context_ptr, &node_options);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    ret = rcl_shutdown(this->old_context_ptr);  // after this, the old_node_ptr should be invalid
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+    this->context_ptr = new rcl_context_t;
+    *this->context_ptr = rcl_get_zero_initialized_context();
+
+    ret = rcl_init(0, nullptr, &init_options, this->context_ptr);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    this->node_ptr = new rcl_node_t;
+    *this->node_ptr = rcl_get_zero_initialized_node();
+    const char * name = "test_graph_node";
+    ret = rcl_node_init(this->node_ptr, name, "", this->context_ptr, &node_options);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+    this->topic_info_array = std::make_unique<rmw_topic_info_array_t>();
```

Comment:
@jaisontj again, why does it have to heap allocated? Also, if used, `std::unique_ptr` should have an associated `Deleter` to properly finalize the allocated `rmw_topic_info_array_t`.

Reply:
Made this member into a value instead of a `unique_ptr`.